pali/COPYING

```diff
@@ -0,0 +1,19 @@
+voc.txt and output.txt are derived from the Digital Pāḷi Dictionary which is licensed under a Creative Commons License: https://digitalpalidictionary.github.io/
+
+To create the files, I downloaded their latest SQL Database and ran the following query:
+
+```sql
+SELECT DISTINCT lemma_2, stem
```

Comment:
@bdhrs - Does this look reasonable to you? :pray:

Reply:
What are you trying to achieve? A little more context would help.

Reply:
Yes of course, Bhante.

SuttaCentral uses ArangoDB to store and search their Pāḷi texts. I'm hoping to implement a stemmer so that searching for a Pāḷi term (e.g. *dhamma*) will return other forms of that word (e.g. *dhamme*).  This is usually accomplished through "stemming" the terms down to some base form that will match (e.g. *dhamm*).

How should I go about using DPD's DB to generate a large set of word to stem mappings in order to generate a test set for such a stemming algorithm?

Reply:
I would avoid stemmers entirely and just use the pre-generated inflections. Your 97.7% accuracy is defintely calcuated by someone who doesn't know Pāḷi, accepting _any_ hit instead of the _correct_ hit. 

In the db, in the 'dpd_headwords' table, there's a column called 'inflections', which already has all the inflected forms of every word, saved as comma-separated values.
![image](https://github.com/snowballstem/snowball-data/assets/64521731/518e8001-e89f-446f-a262-a278aa7628d4)

I assign every word an inflection 'stem' and 'pattern' and generates the entire tables of inflection possibilities, precisely to avoid the messy guesswork of working backwards with stemmers.  

I you want to find which headword any inflected form belongs to, the 'lookup' table provides all that information. That is a json list of 'ID's to be found in the 'dpd_headwords' table.
![image](https://github.com/snowballstem/snowball-data/assets/64521731/ac341777-f65b-411c-89cd-7e48b9bdbab8)

Reply:
@bdhrs - Thanks so much for taking the time to reply!  My responses line by line:

> I would avoid stemmers entirely and just use the pre-generated inflections.

So you're suggesting that instead of using ArangoDB's stemming and normalizing code with a custom Pali stemmer, that instead SuttaCentral rolls its own search indexer in Arango using the DPD's inflections dictionary?  :thinking:  I mean, yes that would  achieve a better accuracy rate. :thinking:  But it also sounds like a larger amount of work (no?). I think for something like fuzzy searching, the stemming doesn't need to be 100% perfect: it just needs to do a better job than the usual regexes which are currently standard when searching the Pali Canon.  Do you disagree?

> Your 97.7% accuracy is defintely calcuated by someone who doesn't know Pāḷi, accepting any hit instead of the correct hit.

Yes, this test set is definitely not correct. That's why I asked you what I should do instead :smile_cat: 

> a column called 'inflections', which already has all the inflected forms of every word, saved as comma-separated values

Ah, thank you! That's what I was looking for. I don't know how I missed that the first time :sweat_smile:  (I think I stopped tabbing over to additional columns when I got to the created_at and updated_at cols, assuming those were the last in the table... :sweat: )

So! Pulling all those `inflections` as the `voc.txt` words and then mapping them to the `stem` column should be the right way to go? :pray:

> avoid the messy guesswork of working backwards with stemmers

I'm curious to hear @ojwb's take on this:  if we already have a large database of every possible inflection, should I even bother adding a (imperfect) stemmer to Snowball?

Reply:
I don't think it's a question I can just answer "yes" or "no" to for you, but I can try to set out some pros and cons to help you decide.

If you have a complete list of every possible inflection of every possible root word in a language then (assuming no inflected forms are duplicated across different root words) you could implement a "perfect" stemmer as a simple look up in a hash table or similar structure.  However I'll point out that stemming is inherently imprecise as it's literally lowering precision to achieve higher recall (https://en.wikipedia.org/wiki/Precision_and_recall), so "perfect" probably sounds more important than it really is in practice.

Your reverse mapping from known inflected forms would be characterised as a dictionary stemmer, whereas Snowball's stemmers are algorithmic.  There are pros and cons to the two approaches - algorithmic stemmers will generally be much smaller (code + data).  Which is faster is less obvious as a dictionary stemmer can use a fairly simple look-up, but an algorithmic stemmer can be much more cache-friendly and speed of information retrieval systems tends to end up cache limited on modern machines.  Algorithmic stemmers also vary a lot in complexity (largely based on the structure of the language being stemmed).

One big advantage of an algorithmic stemmer is it can handle entirely new words and non-dictionary forms of existing words, assuming they follow the pattern of existing words (which they usually do).  It seems from a quick skim of https://en.wikipedia.org/wiki/Pali_literature that Pali is mainly (or maybe entirely?) a historic language so this may not be very relevant in this case.

Adding a stemmer to Snowball has the advantage that anything using Snowball will automatically gain support for that language, so e.g. Xapian, Lucene, Postgres, etc would all gain Pali stemming.  If you implement a custom dictionary-based stemmer in your own system, only your own system gains that support.  If wider support for stemming Pali is something you'd like to see, adding a stemmer to Snowball achieves that as a bonus.

It also occurs to me that given a complete set of inflected forms and stems it might be possible to automatically (or semi-automatically) infer a set of rules which achieve that, and so get an algorithmic stemmer which gives the exact same results (this might be an interesting tool for helping with the creation of new Snowball stemmers more generally as it could be used even with an incomplete set of inflected forms, probably followed by manual adjustments).  Without more knowledge of Pali I'm really not sure how large a set of rules you'd get from this though - in the extreme case where it needed one rule per inflected form that wouldn't be sensible.

Reply:
@ojwb Thanks for your well reasoned response. 

In the course of trying to achieve digital word recognition for the [Digital Pāḷi Dictionary](https://digitalpalidictionary.github.io/), I started with the algorithmic approach, which got me so far, but no further. The "so far" is actually quite good, I estimate about 70% word recognition in early texts, enough for any beginner reader of the language, who reads Pāḷi alongside English translations of Buddhist texts. But it was the "no further" part that troubled me. That 70% recognition drops off steeply to about 20% in later texts. 

The main obstacles are, as you mentioned above, an excessive number of false positives, massive compounding in later texts, and the linguistic feature of adding an indeclinable particle to the end of words, effectively obscuring the inflected form. For example _ka_ is the stem of the pronoun "who, which", _kaṃ_ is an accusative inflected form, _ca_ is an indeclinable particle "and", _kañca_ "and which" is what you find in texts (_kaṃ + ca_ with sandhi). 

These features are present in other languages like Sanskrit as well, I would be interested to know how you deal with them algorithmically?

After reaching a dead-end with the backwards approach of stemming, I currently use the forwards approach of 
- starting with a dictionary lemma
- assigning it an inflection stem + pattern
- generating all possible inflections
- mapping all inflected words in texts to a dictionary lemma(s)

What's unmatched are compounds and sandhi compounds, which I deal with by
- taking a set of all known words and inflected forms, 
- and a set of all known sandhi rules
- simmer those together on a cloud server for 24 hours until separated. 

This approach is much more data intensive, but the accuracy of the results speak for themselves. Of course, this approach is only really possible cos Pāḷi's an effectively dead canonical language, the [corpus of Pāḷi texts](https://tipitaka.org/romn/) is fixed in size, and even though new texts occasionally appear out of Buddhist countries in South East Asia, the vocabulary does not grow, only new derivatives and compounds of already known words.

Even though I don't believe algorithmic stemming is the most effective approach for digital recognition of Pāḷi, as a by-product of my work I have all the data you might need to build the 'best possible stemmer' for the language, and all the Python scripts to process every aspect of the language.

Reply:
> These features are present in other languages like Sanskrit as well, I would be interested to know how you deal with them algorithmically?

Compounding would probably typically be dealt with by using something like ICU to identify word boundaries as the text gets split into words before stemming.

If it was only a matter of dealing with things like `ca`, you might have a first step in the stemming algorithm to remove them (and adjust the remaining stem if necessary, so e.g. `kañca` -> `kaṃ` - the `ca` is effectively ignored by doing this, but "and" is typically ignored anyway as a stop word).  Then a step to remove case endings could remove `ṃ` to leave `ka`.

(For the particular case of a compound word meaning "of which", both of the words comprising it would likely be treated as stopwords anyway.)